presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryBatchStreamReader.java

```diff
@@ -183,16 +182,11 @@ private void openRowGroup()
     {
         // read the dictionary
         if (!dictionaryOpen && dictionarySize > 0) {
-            if (dictionary.length < dictionarySize) {
-                dictionary = new long[dictionarySize];
+            long[] newBuffer = dictionaryProvider.getDictionary(streamDescriptor, dictionary, dictionarySize);
+            if (newBuffer != dictionary) {
```

Comment:
Allocation and memory accounting is complicated here. I would recommend the following.

getDictionary - validate that dictionary has enough size to hold the dictionarySize.
caller - allocate required memory and pass to the getDictionary.

Reply:
I didn't choose that model because otherwise the caller would have to know whether dictionary buffer is going to be allocated at all for the shared dictionary case in the first place. This requires making another call ahead of time, which I don't think is ideal.


Current model is this, caller just passes in the buffer, and dictionary loader decides whether the dictionary buffer is big enough, and return a new one if the buffer is expanded or whether the dictionary buffer is shared. For non-shared dictionary, there is no change to existing accounting, but shared dictionary, this model allows us to more naturally only carry one copy of the dictionary data. I think I also called it out in the google doc and we can discuss more there if required. What do you guys think?